### DISCUSSION TITLE
Programmatically add `order by`

### DISCUSSION THREAD

**Discussion by gmhafiz** [aarondl/sqlboiler#979] [Q&A] [answered]
I would like to add order by clause programmatically. What I have so far is by creating a slice of qm.QueryMod and then looping through a user defined map to sort. The `sortBy` is parsed from url query. Something like this:`http://localhost:8080/api/books?description=t&size=1&page=1&sort=description,asc&sort=created_at,desc`

```go
sortBy := map[string]string{
	"description": "asc",
	"created_at":  "desc",
}

var mods []qm.QueryMod
mods = append(mods, qm.Where("description LIKE ?", "%"+description+"%"))
mods = append(mods, qm.Limit(limit))
mods = append(mods, qm.Offset(offset))

for col, order := range sortBy {
	mods = append(mods, qm.OrderBy(fmt.Sprintf("%s %s", col, order)))
}

query := models.Books(mods...)

res, err := query.All(ctx, r.db)
if err != nil {
	return nil, err
}
```

I get the intended sql query
```sql
SELECT *
FROM `books`
WHERE (description LIKE '%t%')
  AND (`books`.`deleted_at` is null)
ORDER BY description asc, created_at desc
LIMIT 1 OFFSET 1;
```

Is there a better way to construct the query using query mod? Or is there a helper that can add `sortBy` to a normal query builder to this
```go
res, err := models.Books(
    qm.Where("description LIKE ?", "%"+description+"%"),
    qm.Limit(limit),
    qm.Offset(offset),
).
	All(ctx, r.db)
```

**Comment by aarondl** [comment, marked as answer]
There's nothing better than this. Though it doesn't seem like this is too onerous to make a little function in your code that does `OrderByFromMap` since all the types are generic. You can re-assign to mods and the function can change it if it's necessary or return the input if not.

```go
mods := []qm.QueryMod{
  qm.Where("description LIKE ?", "%"+description+"%"),
  qm.Limit(limit),
  qm.Offset(offset),
}

mods = OrderByFromMap(map)
```

**Reply by gmhafiz** [reply]
Hi, it doesn't bother me anymore. Thank you for replying.
